lib/ttfunk/table/cff/charset.rb

```diff
@@ -0,0 +1,230 @@
+                        else
+                          # for cases when there are no sequences at all
+                          Float::INFINITY
+                        end
+
+          # calculate whether storing the charset as a series of ranges is
+          # more efficient (i.e. takes up less space) vs storing it as an
+          # array of SID values
+          total_range_size = (2 * ranges.size) + (range_bytes * ranges.size)
+          total_array_size = sids.size * element_width(:array_format)
+
+          [].tap do |result|
+            if total_array_size <= total_range_size
+              result << [format_int(:array_format)].pack('C')
+              result << sids.pack('n*')
+            else
+              fmt = range_bytes == 1 ? :range_format_8 : :range_format_16
+              element_fmt = element_format(fmt)
+              result << [format_int(fmt)].pack('C')
+
+              ranges.each do |range|
+                sid, num_left = range
+                result << [sid, num_left].pack(element_fmt)
+              end
+            end
+          end.join
+        end
+
+        private
+
+        def sid_for(glyph_id)
+          return 0 if glyph_id == 0
+
+          # rather than validating the glyph as part of one of the predefined
+          # charsets, just pass it through
+          return glyph_id unless offset
+
+          case format_sym
+          when :array_format
+            # zero is always .notdef, so adjust with - 1
+            @entries[glyph_id - 1]
+
+          when :range_format_8, :range_format_16
+            remaining = glyph_id
+
+            @entries.each do |range|
+              if range.size >= remaining
+                return (range.first + remaining) - 1
+              end
+
+              remaining -= range.size
+            end
+
+            0
+          end
+        end
+
+        def find_string(sid)
+          if offset
+            return self.class.standard_strings[sid - 1] if sid <= 390
+
+            idx = sid - 390
+
+            if idx < file.cff.string_index.count
+              file.cff.string_index[idx - 1]
+            end
+          else
+            self.class.strings_for_charset_id(offset_or_id)[sid - 1]
+          end
+        end
+
+        def parse!
+          return unless offset
+          @format = read(1, 'C').first
+
+          case format_sym
+          when :array_format
+            @count = top_dict.charstrings_index.count - 1
+            @length = @count * element_width
+            @entries = read(length, 'n*')
+
+          when :range_format_8, :range_format_16
+            # The number of ranges is not explicitly specified in the font.
+            # Instead, software utilizing this data simply processes ranges
+            # until all glyphs in the font are covered.
+            @count = 0
+            @entries = []
+            @length = 0
+
+            until @count >= top_dict.charstrings_index.count - 1
```

Comment:
I have long wondered why this isn't a pattern already present in TTFunk. Instead all we really have is the `Reader` module which exposes `read` and friends. It would be really nice to be able to define methods like `read_ushort` and give the object some start position and max length. I would think that's something we could file an issue for and implement as a separate chunk of work since it would require a fairly large effort to put into practice across the whole library.